README
WWW::UsePerl::Journal
=====================

DESCRIPTION

An all round journal tool for use.perl addicts. Will access journal entries
for a specific user, or the latest 30 postings, or retrieve a specific
journal entry. Can also post into a specific user's journal.

DEPENDENCIES

The distribution requires the following modules:

  LWP::UserAgent        => 0
  Time::Piece           => 1.07
  Test::More            => 0.47

CHANGES

See the Changes file for a list of updates to the distribution.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

SUPPORT

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please submit a bug to the RT system (see link below). However,
it would help greatly if you are able to pinpoint problems or even supply a 
patch. 

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me by sending an email
to barbie@cpan.org .

RT: http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-UsePerl-Journal

AUTHOR

Original author was Russell Matbouli 
<www-useperl-journal-spam@russell.matbouli.org>, 
<http://russell.matbouli.org/>

Current maintainer is Barbie <barbie@cpan.org>.

COPYRIGHT AND LICENSE

Copyright (C) 2002-2004 Russell Matbouli.
Copyright (C) 2005-2012 Barbie for Miss Barbell Productions.

Distributed under GPL v2. See COPYING included with this distibution.



Changes
Revision history for Perl extension WWW:UsePerl::Journal.

0.25	2012-08-31
	- created connected() method.
	- removed PingTest.pm test library.
	- removed Carp as a prerequisite.

0.24	2012-08-26
	- use.perl.org is now read only. 
	- removed Post.pm features & tests
	- rewritten to work with current URLs
        - removed License files from distro.
        - Artistic License v2 now sole license.
        - spelling fixes. (thanks to Nicholas Bamber, Florian Schlichting & 
          Gregor Herrmann).
        - added minimum perl version (5.006).
        - reworked Makefile.PL for clarity.
	- implemented Perl::Critic suggestions.
	- added META.json
	- change file dates changed to meet W3CDTF standards.

0.23	2009-11-06
	- added PingTest to test suite in order to check whether a network 
	  connection is available.

0.22	2008-09-29
	- patch by Dave Rolsky to fix the login code, to match recent updates 
	  to use.perl.org. (RT#38998 - thanks Dave).
	- updated optional tests.
	- closed security risk in entrytitled (RT#37177 - thanks Yanick)

0.21	2007-11-07
	- updated date testing to try and capture problems CPAN testers are
	  seeing.
	- updated an entry test in the event no content is returned.

0.20	2007-11-07
	- username or userid MUST be provided to new().
	- added debug() method to turn internal debugging on or off.
	- internal function _journalsearch_content renamed to _recent_content().
	- added debug testing script.

0.19	2007-11-05
	- fixed user names with other characters in aside from \w.
	- rewrote the the recent list testing, as posts can appear between
	  extracts.
	- missing entries can have to 2 different error messages, depending
	  upon context.
	- added futher diagnostics in the event of errors.
	- reversed login checks, now assumes failure unless validated.

0.18	2007-11-04
	- changed constants to variable due to suprious errors from 
	  Test::Pod::Coverage.
	- added raw() functions for debugging purposes.
	- changed date processing slightly, hopefully this finally fixes those
	  errors. (thanks to Andreas and Slaven for the feedback)

0.17	2007-10-31
	- Still getting CPAN Tester failures :( It looks like use.perl is
	  timing out, as it's unable to reference the test users (russell and 
	  147 (koschei)). The tests now skip, reporting the error, so I can
	  keep an eye on it.
	- More error reporting in tests if results not as expected, so that
	  CPAN Testers reports can hopefully identify specific problems.

0.16	2007-03-14
	- relaxed date testing, as timezones can affect the date/time value
	  returned.

0.15	2007-03-05
	- fixed several test bugs, including the date parsing issues.
	- WUJ::Post->new() now takes a hash as it states in the docs.
	- Rewrote some of the internals so they correctly cache results. 
	- Bit of an overhaul of the code so that underlying Entry objects are 
	  returned more appropriate, with the caller able to access the Entry
	  methods more directly. 
	- Error handling rewritten so the modules don't fail under pressure,
	  just return undef and set an $j->error().
	- Post.pm has had a fresh look and form completion has been updated.
	- Lots more testing, including developer only tests (not included).
	- Handwritten META.yml to conform to current specs.
	- Changed all instances of LICENCE to LICENSE.
	- Added META.yml test script.
	- Added LICENSE file.
	- Renamed the contrib directory to examples.
	- POD/README updates.
	- Devel::Cover - 93.3% (including developer only tests)

0.14	2005-09-22
	- updated journal post code (RT #4718)
	- added refresh() method to clear the cache
	- removed 22post.t so use.perl isn't bombarded with testing :)

0.13	2005-07-22
	- use.perl layout changes (inc Simon Wilcox - RT #13748)
	- refactored test scripts
	- refactored journal post code
	- Added more test scripts

0.12	2005-04-05
	- New maintainer - Barbie
	- Added sort patch from many eons ago
	- Updated tests to reflect changes to use.perl
	- Added more test scripts

0.11	2003-01-24T15:11:53+00:00
	- Fix for change in the title HTML
	- Date code is still broken - it is out by the timezone of the user
	  who wrote the entry, this may not be fixable if we cannot
	  determine that user's timezone.

0.10	2002-10-14T10:51:25+01:00
	- Patch from Kate L Pugh fixes noon - 1pm bug and newline title bug

0.09	2002-10-09T15:11:57+01:00
	- Patch from Kate L Pugh fixes recentarray regex bug
	- Patch from Richard Clamp to fix his surname and strptime on Mac OS X
	- Contributed scripts for various statistics by 'jdavidb'

0.08	2002-10-04T15:19:30+01:00
	- Change in u.p.o broke recentarray - fixed

0.07	2002-09-27T17:42:51+01:00
	- Patch from Spoon
	- Gets date information from posts
	- "It produces potentially weird results due to time zones. But I believe
	  they are consistently weird. Will probably break if the user logs in."
	  -- the patch author

0.06	2002-09-26T10:46:16+01:00
	- Internal change to use.perl.org broke things
	- Thanks to Simon Wilcox and Simon Wistow for both reporting it on
	  the same day (which confused me ...).

0.05	2002-03-03T20:14:29+00:00
	- Fixes
	- Interface had changed due to the introduction of the Entry class
	- stringification added to it to make it work as it did before
	- date added to Entry class
	- added Test::Simple dependency to makefile.
	- Many thanks to Richard!

0.04	2002-03-03T18:01:36+00:00
	- Changes from koschei
	- WWW::UsePerl::Journal::Entry class
	- recentarray, recent journal entries.

0.03	2002-02-13T12:46:37+00:00
	- Installed Richard Clamp's patches (finally!)
	- Journal list changed on use.perl
	- user id may be supplied instead of username
	- Added LWP dependency to makefile.
	- Thanks Richard!

0.02	2002-01-28T02:16:46+00:00
	- use.perl.org changed some things and everything broke, now fixed.
	- Interface changed, more OO now...
	- Posting possible via postEntry (thanks to Iain Truskett!)
	- Test suite corrected (thanks to Tony Bowden)

0.01	2002-01-23T18:56:32+00:00
	- original version; created by h2xs 1.19


examples/journalstat.pl
#!/usr/local/bin/perl -w
###############################################################################
# Script from http://use.perl.org/~jdavidb/journal/8055 
# modified by russell
###############################################################################

use strict;
use WWW::UsePerl::Journal;

print "Supply usernames to look up on the command line\n"
    if (~~@ARGV == 0);

foreach my $user (@ARGV)
{
    my $journal = WWW::UsePerl::Journal->new($user);
    my @entries = $journal->entryids();

    # Originally I took the date of the first and last entries, but
    # actually I want the current date as an endpoint.  (If you stop
    # posting, that means your average rate should gradually decrease
    # as time progresses.)
    my $firstdate = $journal->entry($entries[0])->date;
    my $numentries = scalar @entries;

    use Time::Piece;
    my $lastdate = localtime;
    my $interval = $lastdate - $firstdate;

    my $per_day = $numentries / $interval->days;

    print "$user has written $per_day entries per day\n";
}


examples/journalmonths.pl
#!/usr/local/bin/perl -w
###############################################################################
# Script from http://use.perl.org/~jdavidb/journal/8055
# modified by russell
# modified by barbie (March 2007)
###############################################################################

use strict;
use WWW::UsePerl::Journal;

print "Supply usernames to look up on the command line\n"
    if (~~@ARGV == 0);

foreach my $user (@ARGV)
{
    my $journal = WWW::UsePerl::Journal->new($user);
    my %entries = $journal->entryhash;
    my %count;

    foreach my $entrynum (sort keys %entries)
    {
        my $entry = $entries{$entrynum};
        my $date = $entry->date;
        my($month, $year) = ($date->mon, $date->year);
        $month = sprintf "%02d", $month;
        $count{"$year$month"}++;
    }

    print "$_:\t$count{$_}\n"   for(sort keys %count)
}


lib/WWW/UsePerl/Journal/Post.pm
package WWW::UsePerl::Journal::Post;

use strict;
use warnings;

use vars qw($VERSION);
$VERSION = '0.23';

#----------------------------------------------------------------------------

=head1 NAME

WWW::UsePerl::Journal::Post - use.perl.org journal post tool

=head1 DESCRIPTION

Allows users to post a use.perl entry to their journal.

=head1 SYNOPSIS

Note that this module is not meant to be called directly. You should access the
undelying object as follows:

  use WWW::UsePerl::Journal;
  my $j = WWW::UsePerl::Journal->new($user);
  my $post = $j->login($pass);

However, you can access the object directly as follows:

  use WWW::UsePerl::Journal;
  my $j = WWW::UsePerl::Journal->new($user);

  use WWW::UsePerl::Journal::Post;
  my $post = WWW::UsePerl::Journal::Post->new(
    j        => $j,
    username => $user,
    password => $pass
  );

  # basic post using defaults
  my $success = $post->postentry(title => 'title', text => 'text');

  # post entry overriding defaults
  $success = $post->postentry(
    title    => 'title',
    text     => 'text',
    topic    => 34,
    comments => 1,
    type     => 1
    promote  => 'publish',
  );

  # delete an entry you don't want
  # WARNING: this is permanent!
  $success = $post->deleteentry($eid);

=cut

# -------------------------------------
# Library Modules

use LWP::UserAgent;
use HTTP::Cookies;
use HTTP::Request::Common;
use Carp;

# -------------------------------------
# Constants & Variables

my $UP_URL = 'http://use.perl.org';

my %postdefaults = (
    topic    => 34,
    comments => 1,
    type     => 1,
    promote  => 'publish',
);

# -------------------------------------
# The Public Interface

=head1 METHODS

=head2 new

Creates an instance for the specified user.

=cut

sub new {
    my $class = shift;
    my %hash  = @_;

    for(qw/j username password/) {
    	return	unless($hash{$_});
    }

    my $ua    = LWP::UserAgent->new(env_proxy => 1);
    $ua->cookie_jar(HTTP::Cookies->new());

    $hash{ua} = $ua;

    my $self  = bless { %hash }, $class;
    return	unless($self->login);
    return $self;
}

sub DESTROY {}

=head2 login

Performs the login.

=cut

sub login {
    my $self = shift;

    my $login =
        $self->{ua}->request(
           GET $UP_URL . "/users.pl?op=userlogin&unickname=$self->{username}&upasswd=$self->{password}"
        )->content;

    return 1    if($login && $login =~ m{\QThis is <b>your</b> User Info page}i);
    return;
}

=head2 postentry

Posts an entry into a journal, given a title and the text of the entry

  $j->postentry(title => "My journal is great", text => "It really is");

=cut

sub postentry {
    my $self = shift;
    my %params = (@_);

    # Validate parameters
    for (qw/text title/) {
        return $self->{j}->error("missing $_ field from postentry")
            unless exists $params{$_};
    }

    %params = (%postdefaults, %params);
    $params{comments} = $params{comments} ? 'enabled' : 'disabled';

    return $self->{j}->error("Invalid post type.")
        if ($params{type}  =~ /[^\d]/ || $params{type}  <  1 ||
                                         $params{type}  >  4);
    return $self->{j}->error("Invalid journal topic.")
        if ($params{topic} =~ /[^\d]/ || $params{topic} <  1  ||
                                         $params{topic} == 16 ||
                                         ($params{topic} > 37 && $params{topic} != 44));
    return $self->{j}->error("Invalid promotion type.")
        if ($params{promote} !~ /(publish|publicize|post)/);
    return $self->{j}->error("Subject too long.")
        if (length($params{title}) > 60);

    # Post posting
    my $editwindow =
        $self->{ua}->request(
        GET $UP_URL . '/journal.pl?op=edit')->content;
    return $self->{j}->error("Can't get an editwindow") unless $editwindow;

    my ($formkey) = ($editwindow =~ m/reskey"\s+value="([^"]+)"/ism);
    return $self->{j}->error("No formkey!") unless defined $formkey;

#promotetype" value="publish|publicize|post

    my %data = (
        id              => '',
        state           => 'editing',
        preview         => 'active',
        reskey          => $formkey,
        description     => $params{title},
        tid             => $params{topic},
        journal_discuss => $params{comments},
        article         => $params{text},
        posttype        => $params{type},
        promotetype     => $params{promote},
        op              => 'save',
    );

    sleep 5;
    my $post = $self->{ua}->request(POST $UP_URL . '/journal.pl', \%data);

    return $self->{j}->error("Failed LWP POST request") unless $post->is_success;
    return $self->{j}->error("Got the wait a little bit message") if $post->content =~ /wait a little bit/;

    return 1;
}

=head2 deleteentry

Deletes an entry from a journal, given the entry id.

  $j->deleteentry($eid);

NOTE: This currently doesn't work!

=cut

sub deleteentry {
    my ($self,$eid) = @_;

    my $content =
        $self->{ua}->request(
        GET $UP_URL . '/journal.pl?op=removemeta&id=' . $eid)->content;
    return $self->{j}->error("No response for deletion") unless $content;

    my ($formkey) = ($content =~ m/reskey"\s+value="([^"]+)"/ism);
    return $self->{j}->error("Unable to delete entry.") unless defined $formkey;

    my %data = (
        reskey => $formkey,
        op     => 'remove',
    );
    $data{'del_'.$eid} = 1;

    my $post = $self->{ua}->request(POST $UP_URL . '/journal.pl', \%data);
    return $self->{j}->error("No response for deletion") unless $post->is_success;

    # Note:
    # I was trying to verify that an entry was deleted, but the code below
    # results in the entry coming back!

#    $content = $post->content;
#    print STDERR "\n#content=\n$content\n";
#    return $self->{j}->error("Jounal entries found!")
#        unless($content =~ /You have not created any journal entries.|Sorry, the requested journal entries were not found./);

    return 1;
}

1;
__END__

=head1 SUPPORT

If you spot a bug or are experiencing difficulties that are not explained
within the POD documentation, please submit a bug to the RT system (see link
below). However, it would help greatly if you are able to pinpoint problems or
even supply a patch.

Fixes are dependant upon their severity and my availablity. Should a fix not
be forthcoming, please feel free to (politely) remind me by sending an email
to barbie@cpan.org .

RT: L<http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-UsePerl-Journal>

=head2 Known Bugs

=over 4

=item * deleteentry

It appears that although the request to delete an entry is correct, there
is obviously something I'm missing, because it doesn't actually get deleted
from the use.perl system :(

=back

=head1 SEE ALSO

L<WWW::UsePerl::Journal>,
F<http://use.perl.org/>

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2003-2009 Barbie for Miss Barbell Productions.

  This module is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

The full text of the licenses can be found in the F<Artistic> and
F<COPYING> files included with this module, or in L<perlartistic> and
L<perlgpl> in Perl 5.8.1 or later.

=cut


lib/WWW/UsePerl/Journal/Entry.pm
package WWW::UsePerl::Journal::Entry;

use strict;
use warnings;

use vars qw($VERSION);
$VERSION = '0.25';

#----------------------------------------------------------------------------

=head1 NAME

WWW::UsePerl::Journal::Entry - use.perl.org journal entry

=head1 DESCRIPTION

Do not use directly. See L<WWW::UsePerl::Journal> for details of usage.

=cut

# -------------------------------------
# Library Modules

use base qw(Class::Accessor::Fast);

use HTTP::Cookies;
use HTTP::Request::Common;
use LWP::UserAgent;
use Time::Piece;
use Time::Seconds;

use WWW::UsePerl::Journal;

#----------------------------------------------------------------------------
# Accessors

=head2 The Accessors

The following accessor methods are available:

  date
  subject
  author
  uid
  content

All functions can be called to return the current value of the associated
object variable.

=cut

__PACKAGE__->mk_accessors($_) for qw(date subject author eid);

# -------------------------------------
# Constants & Variables

my $UP_URL = 'http://use.perl.org/use.perl.org';
use overload q{""}  => sub { $_[0]->stringify() };

my $UID = '
            <div \s+ class="title" \s+ id="user-info-title"> \s+
            <h4> \s+ (.*?) \s+ \((\d+)\) \s+ </h4> \s+ </div>
        ';

my %mons = (
	1  => 'January',
	2  => 'February',
	3  => 'March',
	4  => 'April',
	5  => 'May',
	6  => 'June',
	7  => 'July',
	8  => 'August',
	9  => 'September',
	10 => 'October',
	11 => 'November',
	12 => 'December',
);

# -------------------------------------
# The Public Interface

=head1 INTERFACE

=head2 Constructor

=over 4

=item * new

  use WWW::UsePerl::Journal::Entry;
  my $j = WWW::UsePerl::Journal::Entry->new(%hash);

Creates an instance for a specific entry. The hash must contain values for
the keys 'j' (journal object), 'author' (entry author) and 'eid' (entry id).

=back

=cut

sub new {
    my $class = shift;
    my %opts = (@_);

    for(qw/j author eid/) {
    	return	unless($opts{$_});
    }

    die "No parent object"
	    unless $opts{j}->isa('WWW::UsePerl::Journal');

#use Data::Dumper;
#print STDERR "\n#self->new: ".Dumper(\%opts);

    my $self = bless {%opts}, $class;
    return $self;
}

sub DESTROY {}

=head2 Methods

=over 4

=item * stringify

  use WWW::UsePerl::Journal::Entry;
  my $j = WWW::UsePerl::Journal::Entry->new(%hash);
  print "$j";

Returns the content of the journal entry when the object is directly referenced
in a string.

=cut

sub stringify {
    my $self = shift;
    $self->content();
}

=item * eid

Returns the entry id for the current journal entry.

=cut

sub eid {
    my $self = shift;
    return $self->{eid};
}

=item * content

Return the content of an journal entry.

=cut

sub content {
    my $self   = shift;
    $self->{content} ||= do { $self->_get_content };
}

=item * raw

For debugging purposes.

=back

=cut

sub raw {
    my $self   = shift;
    my $eid    = $self->{eid};
    my $author = $self->{author};
#print STDERR "\n#raw: URL=[". $UP_URL . "/_$author/journal/$eid.html]";
    return $self->{j}->{ua}->request(GET $UP_URL . "/_$author/journal/$eid.html")->content;
}

# -------------------------------------
# The Private Subs

# name:	_get_content
# args:	self .... the current object
# retv: content text
# desc: Given a uid and journal entry id, will retrieve a specific journal
#       entry and disassemble into component parts. returns the content text

sub _get_content {
    my $self   = shift;
    my $eid    = $self->{eid};
    my $author = $self->{author};
    my $content;

    eval {
        $content = $self->{j}->{ua}->request(GET $UP_URL . "/_$author/journal/$eid.html")->content;
    };

#print STDERR "\n#eval=[$@]\n";

    return $self->{j}->error("error getting entry") if($@);

#print STDERR "\n#e->_get_content: URL=[". $UP_URL . "/_$author/journal/$eid.html]";
#print STDERR "\n#content=[$content]\n";

    return $self->{j}->error("error getting entry") unless $content;
    return $self->{j}->error("error getting entry") if($content =~ m!<b>Error type:</b>\s+\d+!);

    return $self->{j}->error("$eid does not exist")
        if $content =~
        m#Sorry, there are no journal entries
        found for this user.</TD></TR></TABLE><P>#is;
    return $self->{j}->error("$eid does not exist")
        if $content =~ m!Sorry, the requested journal entries were not found.!is;

    ($author,$self->{uid}) = $content =~ m!$UID!six;
#print STDERR "\n#e->_get_content: UID=[". ($self->{uid}) ."]";

    ($self->{subject}) = $content =~ m!
        <div \s+ id="journalslashdot"> .*?
        <div \s+ class="title"> \s+
        <h3> \s* (.*?) \s* </h3>
        !six;

    # date/time fields
    my ($month, $day, $year, $hr, $mi, $amp) = $content =~ m!
        <div \s+ class="journaldate">\w+ \s+ (\w+) \s+ (\d+), \s+ (\d+)</div> .*?
        <div \s+ class="details">(\d+):(\d+) \s+ ([AP]M)</div>
        !six;

    unless($day) {
        (undef,$mi,$hr,$day,$month,$year) = localtime(time());
        $month = $mons{$month};
    }

    # just in case we can't get the time
    if($amp) {
        $hr += 12 if($hr >  12 && $amp eq 'PM');
        $hr = 0   if($hr == 12 && $amp eq 'AM');
    }

    # sometimes Time::Piece can't parse the date :(
    eval {
        $self->{date} = Time::Piece->strptime(
            "$month $day $year ${hr}:$mi",
            '%B %d %Y %H:%M'
        );
    };

    #$self->{date} += 4*ONE_HOUR; # correct TZ?

    $content =~ m! <div \s+ class="intro">\s*(.*?)\s*</div> !six;
    return $1;
}

1;

__END__

=head1 CAVEATS

Beware the stringification of WWW::UsePerl::Journal::Entry objects.
They're still objects, they just happen to look the same as before when
you're printing them. Use -E<gt>content instead.

The time on a journal entry is the localtime of the user that created the
journal entry. If you aren't in the same timezone, that time will be wrong.

=head1 SUPPORT

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please submit a bug to the RT system (see link below). However,
it would help greatly if you are able to pinpoint problems or even supply a
patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me by sending an email
to barbie@cpan.org .

RT: L<http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-UsePerl-Journal>

=head1 SEE ALSO

F<http://use.perl.org/use.perl.org>

L<WWW::UsePerl::Journal::Server>

=head1 AUTHOR

  Original author: Russell Matbouli
  <www-useperl-journal-spam@russell.matbouli.org>,
  <http://russell.matbouli.org/>

  Current maintainer: Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2002-2004 Russell Matbouli.
  Copyright (C) 2005-2012 Barbie for Miss Barbell Productions.

This module is free software; you can redistribute it and/or
modify it under the Artistic Licence v2.

=cut


Makefile.PL
#!/usr/bin/perl

use strict;
use warnings;
use 5.006;

use ExtUtils::MakeMaker;

WriteMakefile(
    AUTHOR              => 'Barbie <barbie@cpan.org>',
    NAME                => 'WWW::UsePerl::Journal',
    VERSION_FROM        => 'lib/WWW/UsePerl/Journal.pm',
    ABSTRACT            => 'Handles the retrieval of UsePerl journal entries',
    NO_META             => 1,
    PREREQ_PM           => {

        # runtime prereqs
        'Carp'                      => 0,
        'Class::Accessor::Fast'     => 0,
        'HTTP::Cookies'             => 0,
        'HTTP::Request::Common'     => 0,
        'LWP::UserAgent'            => 0,
        'Time::Piece'               => 1.07,
        'Time::Seconds'             => 0,

        # build/test prereqs
        'IO::File'                  => 0,
        'Test::Exception'           => 0,
		'Test::More'                => 0.70,

    }
);
